File: frontend/src/api/trace.ts
import api from './client'

export interface TraceSummary {
  trace_id: string
  session_id: string
  status: string
  start_time: string
  duration_ms: number | null
  mode: string
  total_tokens: number
  agents_involved: string[]
  span_count: number
  question_summary: string
}

export interface TraceListResponse {
  traces: TraceSummary[]
  total: number
  limit: number
  offset: number
}

export interface WaterfallSpan {
  id: string
  parent_id: string | null
  span_type: string
  name: string
  status: string
  start_offset_ms: number
  duration_ms: number
  depth: number
  error_message?: string | null
  attributes: Record<string, any>
}

export interface WaterfallResponse {
  trace_id: string
  total_duration_ms: number
  spans: WaterfallSpan[]
}

export interface AgentStats {
  [agentId: string]: {
    call_count: number
    avg_duration_ms: number
    p50_ms: number
    p90_ms: number
    success_rate: number
    avg_tokens: number
  }
}

export interface ToolStats {
  [toolName: string]: {
    call_count: number
    avg_duration_ms: number
    success_rate: number
  }
}

export interface LLMStats {
  call_count: number
  avg_latency_ms: number
  p50_ms: number
  p90_ms: number
  avg_prompt_tokens: number
  avg_completion_tokens: number
  total_prompt_tokens: number
  total_completion_tokens: number
}

export interface StageBreakdown {
  [stageName: string]: number
}

export interface SlowTraceItem {
  trace_id: string
  session_id: string
  duration_ms: number
  mode: string
  agents_involved: string[]
  question_summary: string
}

export interface ErrorTraceItem {
  trace_id: string
  session_id: string
  duration_ms: number
  mode: string
  question_summary: string
  start_time: string
}

export async function getTraces(limit = 50, offset = 0, sessionId?: string) {
  const { data } = await api.get<TraceListResponse>('/traces', {
    params: { limit, offset, session_id: sessionId },
  })
  return data
}

export async function getTraceTree(traceId: string) {
  const { data } = await api.get(`/traces/${traceId}`)
  return data
}

export async function getTraceSpans(traceId: string) {
  const { data } = await api.get(`/traces/${traceId}/spans`)
  return data.spans || []
}

export async function getTraceWaterfall(traceId: string) {
  const { data } = await api.get<WaterfallResponse>(`/traces/${traceId}/waterfall`)
  return data
}

export async function getTraceStages(traceId: string) {
  const { data } = await api.get(`/traces/${traceId}/stages`)
  return data.stages || {}
}

export async function getAgentStats(days = 7) {
  const { data } = await api.get('/traces/stats/agents', { params: { days } })
  return data.stats || {}
}

export async function getToolStats(days = 7) {
  const { data } = await api.get('/traces/stats/tools', { params: { days } })
  return data.stats || {}
}

export async function getLLMStats(days = 7) {
  const { data } = await api.get('/traces/stats/llm', { params: { days } })
  return data.stats || {}
}

export async function getSlowTraces(thresholdMs = 30000, limit = 10) {
  const { data } = await api.get('/traces/stats/slow', {
    params: { threshold_ms: thresholdMs, limit },
  })
  return data.traces || []
}

export async function getErrorTraces(days = 7, limit = 20) {
  const { data } = await api.get('/traces/stats/errors', {
    params: { days, limit },
  })
  return data.traces || []
}

File: mediZJ/api/models/trace.py
"""Trace API 的 Pydantic 响应模型"""
from typing import Dict, Any, Optional, List
from pydantic import BaseModel


class SpanItem(BaseModel):
    """扁平 span 条目"""
    id: str
    trace_id: str
    parent_id: Optional[str] = None
    span_type: str
    name: str = ""
    status: str = "ok"
    start_time: str = ""
    end_time: Optional[str] = None
    duration_ms: Optional[float] = None
    error_message: Optional[str] = None
    llm_attrs: Optional[Dict[str, Any]] = None
    tool_attrs: Optional[Dict[str, Any]] = None
    agent_attrs: Optional[Dict[str, Any]] = None


class TraceSummary(BaseModel):
    """trace 列表汇总条目"""
    trace_id: str
    session_id: str = ""
    status: str = "ok"
    start_time: str = ""
    duration_ms: Optional[float] = None
    mode: str = ""
    total_tokens: int = 0
    agents_involved: List[str] = []
    span_count: int = 0
    question_summary: str = ""


class TraceListResponse(BaseModel):
    """trace 列表响应"""
    traces: List[TraceSummary] = []
    total: int = 0
    limit: int = 50
    offset: int = 0


class WaterfallSpan(BaseModel):
    """Waterfall 视图的 span 条目（含 offset 和 depth）"""
    id: str
    parent_id: Optional[str] = None
    span_type: str
    name: str
    status: str = "ok"
    start_offset_ms: float = 0
    duration_ms: float = 0
    depth: int = 0
    error_message: Optional[str] = None
    attributes: Dict[str, Any] = {}


class WaterfallResponse(BaseModel):
    """Waterfall 视图响应"""
    trace_id: str
    total_duration_ms: float = 0
    spans: List[WaterfallSpan] = []


class AgentStat(BaseModel):
    """per-agent 统计"""
    call_count: int = 0
    avg_duration_ms: float = 0
    p50_ms: float = 0
    p90_ms: float = 0
    success_rate: float = 0
    avg_tokens: int = 0


class ToolStat(BaseModel):
    """per-tool 统计"""
    call_count: int = 0
    avg_duration_ms: float = 0
    success_rate: float = 0


class LLMStat(BaseModel):
    """LLM 调用统计"""
    call_count: int = 0
    avg_latency_ms: float = 0
    p50_ms: float = 0
    p90_ms: float = 0
    avg_prompt_tokens: int = 0
    avg_completion_tokens: int = 0
    total_prompt_tokens: int = 0
    total_completion_tokens: int = 0


class StageBreakdown(BaseModel):
    """阶段耗时分布"""
    trace_id: str
    stages: Dict[str, float] = {}


class SlowTraceItem(BaseModel):
    """慢 trace 条目"""
    trace_id: str
    session_id: str = ""
    duration_ms: float = 0
    mode: str = ""
    agents_involved: List[str] = []
    question_summary: str = ""


class ErrorTraceItem(BaseModel):
    """错误 trace 条目"""
    trace_id: str
    session_id: str = ""
    duration_ms: float = 0
    mode: str = ""
    question_summary: str = ""
    start_time: str = ""